### PR TITLE
day 7 solution

### PR DESCRIPTION
**Pull request by Joshua-Tustanowski**

### CODE REVIEW

File: Josh/solutions_2021/day_7/solution.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    for key in v.keys():
+        if v[key] == min_value:
+            return key, v[key]
+    return -1, -1
+
+
+def part_one(file_name: str) -> Tuple[int, int]:
+    initial_positions = parse_input(file_name)
+    fuel_usages = {}
+    min_pos, max_pos = min(initial_positions), max(initial_positions)
```
Comment (uttamo) [inline Josh/solutions_2021/day_7/solution.py:27]: would be quicker if you sorted `initial_positions` once and then did
```python
min_pos, max_pos = initial_positions[0], initial_positions[-1]
```
?

File: Josh/solutions_2021/day_7/solution.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        fuel_usage = 0
+        for pos in initial_positions:
+            val = abs(pos - value)
+            fuel_usage += sum(values[:val])
+        fuel_usages[value] = fuel_usage
+    return get_minimum_entry(fuel_usages)
+
+
+def maths(file_name: str) -> int:
+    """ See submarine.pdf for a writeup from Reddit u/throwaway7824365346 on why the mean works """
```
Comment (uttamo) [inline Josh/solutions_2021/day_7/solution.py:53]: wow 😂

File: Josh/solutions_2021/day_7/solution.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+CURRENT_DIR = dirname(__file__)
+
+
+def parse_input(file_name: str) -> List[int]:
+    with open(join(CURRENT_DIR, file_name), "r") as fp:
+        contents = fp.read().split(",")
+    return [int(v) for v in contents]
+
+
+def get_minimum_entry(v: Dict[int, int]) -> Tuple[int, int]:
```
Comment (uttamo) [inline Josh/solutions_2021/day_7/solution.py:16]: nice typing

File: Josh/solutions_2021/day_7/solution.py (hunk 1 above)
Comment (Joshua-Tustanowski) [reply Josh/solutions_2021/day_7/solution.py:27]: idk man my current implementation is O(DM) where D is the difference between the min and max positions and M is the number os initial positions. If I sorted it, then the complexity would still be limited by my awful brute force loop

File: Josh/solutions_2021/day_7/solution.py (hunk 2 above)
Comment (Joshua-Tustanowski) [reply Josh/solutions_2021/day_7/solution.py:53]: It's a pretty good writeup 🥇

File: Josh/solutions_2021/day_7/solution.py (hunk 1 above)
Comment (uttamo) [reply Josh/solutions_2021/day_7/solution.py:27]: ah right so my suggestion's performance improvement is basically negligible.

One alternative approach I saw was to start in the middle (mean or median) then
* calculate total fuel usage
* move position 1 to the right or left
* calculate total fuel usage again, which should hopefully be lower
* if the total fuel usage increases, then you've reached the minimum

kind of like gradient descent

Comment (Joshua-Tustanowski) [reply Josh/solutions_2021/day_7/solution.py:27]: Sounds interesting, I was going to try and find a closed form solution which would probably look like gradient descent after applying the method the guy used for finding out the mean was a good solution

**Pull request merged**